exec/java-exec/src/main/java/org/apache/drill/exec/rpc/user/UserClientAuthenticationHandler.java

```diff
@@ -0,0 +1,229 @@
+
+    final SaslMessage challenge;
+    final SaslClient saslClient;
+    final UserGroupInformation ugi;
+    final SettableFuture<Void> settableFuture;
+
+    public SaslChallengeContext(SaslMessage challenge, SaslClient saslClient, UserGroupInformation ugi,
+                                SettableFuture<Void> settableFuture) {
+      this.challenge = checkNotNull(challenge);
+      this.saslClient = checkNotNull(saslClient);
+      this.ugi = checkNotNull(ugi);
+      this.settableFuture = checkNotNull(settableFuture);
+    }
+  }
+
+  private interface SaslChallengeProcessor {
+
+    /**
+     * Process challenge from server, and return a response.
+     *
+     * Returns null iff SASL exchange is complete and successful.
+     *
+     * @param context challenge context
+     * @return response
+     * @throws Exception
+     */
+    SaslMessage process(SaslChallengeContext context) throws Exception;
+
+  }
+
+  private static class SaslInProgressProcessor implements SaslChallengeProcessor {
+
+    @Override
+    public SaslMessage process(SaslChallengeContext context) throws Exception {
+      final SaslMessage.Builder response = SaslMessage.newBuilder();
+
+      final byte[] responseBytes = evaluateChallenge(context.ugi, context.saslClient,
+          context.challenge.getData().toByteArray());
+
+      final boolean isComplete = context.saslClient.isComplete();
+      logger.trace("Evaluated challenge. Completed? {}.", isComplete);
+      response.setData(responseBytes != null ? ByteString.copyFrom(responseBytes) : ByteString.EMPTY);
+      // if isComplete, the client will get one more response from server
+      response.setStatus(isComplete ? SaslStatus.SASL_SUCCESS : SaslStatus.SASL_IN_PROGRESS);
+      return response.build();
+    }
+  }
+
+  private static class SaslSuccessProcessor implements SaslChallengeProcessor {
+
+    @Override
+    public SaslMessage process(SaslChallengeContext context) throws Exception {
+      if (context.saslClient.isComplete()) {
+        logger.trace("Successfully authenticated to server using {}", context.saslClient.getMechanismName());
+        return null;
+      } else {
+
+        // server completed before client; so try once, fail otherwise
+        evaluateChallenge(context.ugi, context.saslClient,
+            context.challenge.getData().toByteArray()); // discard response
+
+        if (context.saslClient.isComplete()) {
+          logger.trace("Successfully authenticated to server using {}", context.saslClient.getMechanismName());
+          return null;
+        } else {
+          throw new SaslException("Server allegedly succeeded authentication, but client did not. Suspicious?");
+        }
+      }
+    }
+  }
+
+  private static class SaslFailedProcessor implements SaslChallengeProcessor {
+
+    @Override
+    public SaslMessage process(SaslChallengeContext context) throws Exception {
+      throw new SaslException("Authentication failed. Incorrect credentials?");
```

Comment:
Typically the client is not notified why auth failed, except that something went wrong.